apps/src/gamelab/AnimationPicker/AnimationPreview.jsx

```diff
@@ -0,0 +1,129 @@
+/** @file Render a gallery image/spritesheet as an animated preview */
+
+var MARGIN_PX = 2;
+
+/**
+ * Render an animated preview of a spritesheet at a given size, scaled with
+ * a fixed aspect ratio to fit.
+ */
+var AnimationPreview = React.createClass({
+  propTypes: {
+    width: React.PropTypes.number.isRequired,
+    height: React.PropTypes.number.isRequired,
+    sourceUrl: React.PropTypes.string.isRequired,
+    sourceWidth: React.PropTypes.number.isRequired,
+    sourceHeight: React.PropTypes.number.isRequired,
+    frameWidth: React.PropTypes.number.isRequired,
+    frameHeight: React.PropTypes.number.isRequired,
+    frameCount: React.PropTypes.number.isRequired,
+    frameRate: React.PropTypes.number.isRequired
+  },
+
+  getInitialState: function () {
+    return {
+      currentFrame: 0,
+      framesPerRow: 1,
+      scaledSourceWidth: 0,
+      scaledFrameWidth: 0,
+      scaledFrameHeight: 0,
+      extraTopMargin: 0,
+      wrappedSourceUrl: ''
+    };
+  },
+
+  componentWillMount: function () {
+    this.precalculateRenderProps(this.props);
+  },
+
+  componentWillReceiveProps: function (nextProps) {
+    this.precalculateRenderProps(nextProps);
+  },
+
+  componentWillUnmount: function () {
+    if (this.timeout_) {
+      clearTimeout(this.timeout_);
+    }
+  },
+
+  onMouseOver: function () {
+    this.advanceFrame();
+  },
+
+  onMouseOut: function () {
+    this.stopAndResetAnimation();
+  },
+
+  advanceFrame: function () {
+    this.setState({ currentFrame: (this.state.currentFrame + 1) % this.props.frameCount });
+    this.timeout_ = setTimeout(this.advanceFrame, 1000 / this.props.frameRate);
+  },
+
+  stopAndResetAnimation: function () {
+    if (this.timeout_) {
+      clearTimeout(this.timeout_);
+      this.timeout_ = undefined;
+    }
+    this.setState({ currentFrame: 0 });
+  },
+
+  precalculateRenderProps: function (nextProps) {
+    var framesPerRow = Math.floor(this.props.sourceWidth / this.props.frameWidth);
+    var innerWidth = nextProps.width - 2 * MARGIN_PX;
+    var innerHeight = nextProps.height - 2 * MARGIN_PX;
+    var xScale = innerWidth / nextProps.frameWidth;
+    var yScale = innerHeight / nextProps.frameHeight;
+    var scale = Math.min(1, Math.min(xScale, yScale));
+    var scaledFrameHeight = nextProps.frameHeight * scale;
+    this.setState({
+      framesPerRow: framesPerRow,
+      scaledSourceWidth: nextProps.sourceWidth * scale,
+      scaledFrameWidth: nextProps.frameWidth * scale,
+      scaledFrameHeight: scaledFrameHeight,
+      extraTopMargin: Math.ceil((innerHeight - scaledFrameHeight) / 2),
+      wrappedSourceUrl: "url('" + this.props.sourceUrl + "')"
+    });
```

Comment:
Couple places in here we use `this.props` - should we actually be using `nextProps`?